core/src/main/java/org/bitcoinj/evolution/SimplifiedMasternodeListManager.java

```diff
@@ -811,18 +1057,416 @@ public SimplifiedQuorumList getQuorumListForBlock(Sha256Hash blockHash) {
+                return null;
             }
-            return null;
+        } catch (BlockStoreException x) {
+            // we are in deep trouble
+            throw new RuntimeException(x);
         } finally {
             lock.unlock();
         }
     }
 
+    private ArrayList<ArrayList<Masternode>> computeQuorumMembersByQuarterRotation(LLMQParameters.LLMQType llmqType, StoredBlock quorumBaseBlock) throws BlockStoreException {
+        final LLMQParameters llmqParameters = LLMQParameters.fromType(llmqType);
+
+        final int cycleLength = llmqParameters.getDkgInterval();
+
+        final StoredBlock blockHMinusC = quorumBaseBlock.getAncestor(blockChain.getBlockStore(), quorumBaseBlock.getHeight() - cycleLength);
+        final StoredBlock pBlockHMinus2C = quorumBaseBlock.getAncestor(blockChain.getBlockStore(),quorumBaseBlock.getHeight() - 2 * cycleLength);
+        final StoredBlock pBlockHMinus3C = quorumBaseBlock.getAncestor(blockChain.getBlockStore(),quorumBaseBlock.getHeight() - 3 * cycleLength);
+
+        PreviousQuorumQuarters previousQuarters = getPreviousQuorumQuarterMembers(llmqParameters, blockHMinusC, pBlockHMinus2C, pBlockHMinus3C);
+
+        ArrayList<ArrayList<Masternode>> quorumMembers = Lists.newArrayListWithCapacity(llmqParameters.getSigningActiveQuorumCount());
+        for (int i = 0; i < llmqParameters.getSigningActiveQuorumCount(); ++i) {
+            quorumMembers.add(Lists.newArrayList());
+        }
+
+        ArrayList<ArrayList<SimplifiedMasternodeListEntry>> newQuarterMembers = buildNewQuorumQuarterMembers(llmqParameters, quorumBaseBlock, previousQuarters);
+
+        for (int i = 0; i < llmqParameters.getSigningActiveQuorumCount(); ++i) {
+            for (SimplifiedMasternodeListEntry m: previousQuarters.quarterHMinus3C.get(i)) {
+                for (int j = 0; j < quorumMembers.get(i).size(); ++j) {
+                    if (m.equals(quorumMembers.get(i).get(j))) {
+                        log.info("{} is already in the list", m);
+                    }
+                }
+                quorumMembers.get(i).add(m);
+            }
+            for (SimplifiedMasternodeListEntry m: previousQuarters.quarterHMinus2C.get(i)) {
+                for (int j = 0; j < quorumMembers.get(i).size(); ++j) {
+                    if (m.equals(quorumMembers.get(i).get(j))) {
+                        log.info("{} is already in the list", m);
+                    }
+                }
+                quorumMembers.get(i).add(m);
+            }
+            for (SimplifiedMasternodeListEntry m: previousQuarters.quarterHMinusC.get(i)) {
+                for (int j = 0; j < quorumMembers.get(i).size(); ++j) {
+                    if (m.equals(quorumMembers.get(i).get(j))) {
+                        log.info("{} is already in the list", m);
+                    }
+                }
+                quorumMembers.get(i).add(m);
+            }
+            for (SimplifiedMasternodeListEntry m: newQuarterMembers.get(i)) {
```

Comment:
Repeating code - might be better to extract into a function.

Reply:
Good idea. part of this code is to check for duplicates in lists (which is a bug I am trying to find the cause of).